sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AesEncryptionShouldBeUsedWithSecuredMode.cs

```diff
@@ -0,0 +1,78 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class AesEncryptionShouldBeUsedWithSecuredMode : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4432";
+        private const string MessageFormat = "Use Galois/Counter Mode (GCM/NoPadding) instead.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<string> invalidEnumValues =
+            new HashSet<string>
+            {
+                "CipherMode.CBC",
+                "CipherMode.CFB",
+                "CipherMode.CTS",
+                "CipherMode.ECB",
+                "CipherMode.OFB"
```

Comment:
It looks like these are all of the cipher modes supported by _AesManaged_ - are there any possible valid uses of the class?

Reply:
There are none and there is no compliant encryption class built-in the framework, you'll have to use third parties.

Reply:
Also, according to [MSDN](https://msdn.microsoft.com/en-us/library/system.security.cryptography.aesmanaged.mode(v=vs.110).aspx) the default value for _Mode_ is _CBC_. However, we're only checking for assignments to _Mode_ so we won't catch default usage.

Is that ok i.e. should we just treat it as a false negative? If so, it might be worth documenting somewhere, either as a comment in the code or in the test cases.

Reply:
I forgot to look for the default value. Maybe the best thing is to report on `AesManaged` creation as there is no secure cipher for the time being. WDYT?

Reply:
It's probably good enough - it will always be correct unless and until a new secure version is included in the framework. We could write a smarter version at that point.